### PR TITLE
2799: Implement chat api

### PR DESCRIPTION
**Pull request by f1sh1918**
### Short description

As a user i want to get information directly from a service user. Now the user can reach a service user that uses zammad help desk

### Proposed changes

- add get and post endpoint for chat
- add polling for message refresh
- add uuid creation for device id

### Side effects

- none
### Resolved issues

Fixes: #2799

### Testing
1. Start a chat (Testumgebung or Landeshauptstadt München)
2. Login to zammad helpdesk :https://zammad-integreat.tuerantuer.org/
3. Use credentials: https://passbolt.tuerantuer.org/app/passwords/view/a9e7cb57-2a92-453b-bb3d-c89fe8bbe1f9
4. Check for the deviceId in `localStorage` and look for the chat in zammad
5. Answer the chat message (message polling is 16s now)

Mind that showing a loading spinner can be problematic, since its reloading the chat with polling. Just showing loading spinner on initial loading

### CODE REVIEW

File: web/src/components/ChatController.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        refreshMessages()
+      }
+    }
+
+    if (error !== null) {
+      setSendingStatus('failed')
+    }
+  }
+
+  if (!hasOpenChatSession) {
```
Comment (LeandraH) [inline web/src/components/ChatController.tsx:48]: 🔧  Can we turn this around to have `if (hasOpenChatSession) { return <ChatInitializedView /> } return <Chat />`?

File: web/src/components/__tests__/Chat.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      deviceId={deviceId}
+      hasError={hasError}
+      isLoading={isLoading}
+      submitMessage={submitMessage}
+      refreshMessages={refreshMessages}
+    />,
+  )
+
+describe('Chat', () => {
+  it('should sending button be disabled if no input text is provided', () => {
```
Comment (LeandraH) [inline web/src/components/__tests__/Chat.spec.tsx:33]:
```suggestion
  it('sending button should be disabled if no input text is provided', () => {
```

File: web/src/components/__tests__/Chat.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    const { getByRole } = render([], false, false, submitMessage, refresh, undefined)
+    const buttonSendMessage = getByRole('button', {
+      name: 'chat:sendButton',
+    })
+    expect(buttonSendMessage).toBeDisabled()
+  })
+  it('should be able to send a message if text is provided', () => {
```
Comment (LeandraH) [inline web/src/components/__tests__/Chat.spec.tsx:40]:
```suggestion

  it('should be able to send a message if text is provided', () => {
```

File: web/src/components/__tests__/ChatContainer.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  it('should open chat modal and show content on chat button click', () => {
+    const { getByTestId, getByText } = renderWithTheme(<ChatContainer city='Augsburg' language='de' />)
+    const chatButtonContainer = getByTestId('chat-button-container')
+    expect(chatButtonContainer).toBeTruthy()
+    fireEvent.click(chatButtonContainer)
+    expect(getByText('chat:header')).toBeTruthy()
+    expect(getByText('chat:conversationTitle')).toBeTruthy()
+    expect(getByText('chat:conversationText')).toBeTruthy()
+  })
+  it('should show minimized toolbar on minimize button click', () => {
```
Comment (LeandraH) [inline web/src/components/__tests__/ChatContainer.spec.tsx:20]:
```suggestion

  it('should show minimized toolbar on minimize button click', () => {
```

File: web/src/components/__tests__/ChatContainer.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    const { getByTestId, getByLabelText } = renderWithTheme(<ChatContainer city='Augsburg' language='de' />)
+    const chatButtonContainer = getByTestId('chat-button-container')
+    expect(chatButtonContainer).toBeTruthy()
+    fireEvent.click(chatButtonContainer)
+    const minimizeButton = getByLabelText('minimize')
+    fireEvent.click(minimizeButton)
+    expect(getByTestId('chat-minimized-toolbar')).toBeTruthy()
+  })
+  it('should close chat if close button was clicked', () => {
```
Comment (LeandraH) [inline web/src/components/__tests__/ChatContainer.spec.tsx:29]:
```suggestion

  it('should close chat if close button was clicked', () => {
```

File: web/src/components/__tests__/ChatContainer.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    expect(chatButtonContainer).toBeTruthy()
+    fireEvent.click(chatButtonContainer)
+    const closeButton = getByLabelText('close')
+    fireEvent.click(closeButton)
+    expect(queryByTestId('chat-minimized-toolbar')).toBeFalsy()
+    expect(queryByText('chat:header')).toBeFalsy()
+    expect(queryByText('chat:conversationTitle')).toBeFalsy()
+    expect(queryByText('chat:conversationText')).toBeFalsy()
+  })
+  it('should maximize chat on maximize button click if chat was minimized', () => {
```
Comment (LeandraH) [inline web/src/components/__tests__/ChatContainer.spec.tsx:43]:
```suggestion

  it('should maximize chat on maximize button click if chat was minimized', () => {
```

File: web/src/components/__tests__/ChatConversation.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    <ChatConversation messages={messages} hasConversationStarted={hasConversationStarted} hasError={hasError} />,
+  )
+
+describe('ChatConversation', () => {
+  it('should display welcome text if conversation has not started', () => {
+    const { getByText } = render([], false, false)
+    expect(getByText('chat:conversationTitle')).toBeTruthy()
+    expect(getByText('chat:conversationText')).toBeTruthy()
+  })
+  it('should display messages if conversation has started and the initial message', () => {
```
Comment (LeandraH) [inline web/src/components/__tests__/ChatConversation.spec.tsx:23]:
```suggestion

  it('should display initial and following messages if conversation was started', () => {
```

File: web/src/components/__tests__/ChatConversation.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    const { getByText, getByTestId } = render(testMessages, true, false)
+    expect(getByText('chat:initialMessage')).toBeTruthy()
+    expect(getByTestId(testMessages[0]!.id)).toBeTruthy()
+    expect(getByTestId(testMessages[1]!.id)).toBeTruthy()
+  })
+  it('should display error messages if error occurs', () => {
```
Comment (LeandraH) [inline web/src/components/__tests__/ChatConversation.spec.tsx:29]:
```suggestion

  it('should display error messages if error occurs', () => {
```

**Comment by LeandraH** [comment]
> Not sure if we should limit the polling somehow. If someone keeps the browser open forever, the polling won't stop

Good point, that's not great. Most modern browsers will probably kill the tab at some point but that's not exactly who we are building for...

**Review by LeandraH — APPROVED** [review]
Looks great and works, thank you! Tested in Firefox.

I did manage to see the small horizontal scrollbar that Steffen was talking about and managed to fix it by setting the `width` and `height` of the `IconContainer` in the `ChatMessage` to `25px` instead of `24px`.

File: shared/api/endpoints/__tests__/createChatMessagesEndpoint.spec.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    language: 'fa',
+    deviceId: '23123-dsasd1-2dsa12',
+  }
+  const endpoint = createChatMessagesEndpoint(baseUrl)
+  it('should map params to url', () => {
+    expect(endpoint.mapParamsToUrl(params)).toBe(
+      `${baseUrl}/api/${API_VERSION}/${params.city}/${params.language}/chat/${params.deviceId}/`,
+    )
+  })
+  it('should map fetched data to model', () => {
```
Comment (LeandraH) [inline shared/api/endpoints/__tests__/createChatMessagesEndpoint.spec.ts:18]:
```suggestion

  it('should map fetched data to model', () => {
```

File: shared/api/endpoints/__tests__/createChatSessionEndpoint.spec.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    deviceId: '23123-dsasd1-2dsa12',
+    message: 'Ich habe eine Frage',
+  }
+  const endpoint = createChatSessionEndpoint(baseUrl)
+  it('should map params to url', () => {
+    expect(endpoint.mapParamsToUrl(params)).toBe(
+      `${baseUrl}/api/${API_VERSION}/${params.city}/${params.language}/chat/${params.deviceId}/`,
+    )
+  })
+  it('should map the params to the body', () => {
```
Comment (LeandraH) [inline shared/api/endpoints/__tests__/createChatSessionEndpoint.spec.ts:18]:
```suggestion

  it('should map the params to the body', () => {
```

**Review by LeandraH — APPROVED** [review]
Nice, thanks for adding more tests, they worked on my machine too 😁 

One small thing: I think the shadow on the chat icon is off-center, at least for me in Firefox.
image

File: shared/api/models/CityModel.ts [hunk 11]
```diff
@@ -79,6 +82,9 @@ class CityModel {
     return this._sortingName
   }
 
+  get chatEnabled(): boolean {
+    return this._chatEnabled
+  }
```
Comment (LeandraH) [inline shared/api/models/CityModel.ts:87]:
```suggestion
  }

```

**Comment by f1sh1918** [comment]
> Nice, thanks for adding more tests, they worked on my machine too 😁
> 
> One small thing: I think the shadow on the chat icon is off-center, at least for me in Firefox. image

what do you mean by "off center" ? it should look like that

**Comment by LeandraH** [comment]
> > Nice, thanks for adding more tests, they worked on my machine too 😁
> > One small thing: I think the shadow on the chat icon is off-center, at least for me in Firefox. 
> 
> what do you mean by "off center" ? it should look like that

Oh, funny. I guess that's fine then. I'd have expected the center of the shadow to also be the center of the icon 🤷

File: web/src/components/ChatController.tsx [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+  if (hasOpenChatSession) {
+    return (
+      <ChatInitializedView deviceId={storedDeviceId} city={city} language={language} submitMessage={submitMessage} />
+    )
+  }
+  return (
+    <Chat
+      submitMessage={submitMessage}
+      messages={[]}
```
Comment (steffenkleinle) [inline web/src/components/ChatController.tsx:56]: Is it on purpose that we pass an empty messages array here?

**Review by steffenkleinle — APPROVED** [review]
Nice work! Tested on firefox, works perfectly!

Obviously not from this PR:
- the content behind the chat layover is scrollable
- on small screens the send button is cut off at the bottom
image

**Comment by f1sh1918** [comment]
@steffenkleinle and @LeandraH i added a hook to check if the browser tab is active and to restrict the polling to active tab. Because if someone opens the chat keeps the tab open for hours without using, the polling would continue.

Please check the last commit

File: web/src/components/ChatController.tsx (hunk 12 above)
Comment (f1sh1918) [reply web/src/components/ChatController.tsx:56]: we could also pass null depends on the checks in the chat component

File: shared/api/endpoints/createChatSessionEndpoint.ts [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+export const CHAT_ENDPOINT_NAME = 'chat'
+type ParamsType = {
+  city: string
+  language: string
+  deviceId: string
+  message: string
+}
+
+export default (baseUrl: string): Endpoint<ParamsType, Array<ChatMessageModel>> =>
+  new EndpointBuilder<ParamsType, Array<ChatMessageModel>>(CHAT_ENDPOINT_NAME)
```
[the remark this answers is not in the document]
Comment (f1sh1918) [reply shared/api/endpoints/createChatSessionEndpoint.ts:15]: yes but the cms named both "chat"
I can adjust the filename

File: translations/translations.json [hunk 14]
```diff
@@ -211,10 +211,12 @@
       "conversationTitle": "Bitte geben Sie Ihre Frage in das Textfeld ein.",
       "conversationText": "Sie können alles fragen, von lokalen Informationen bis hin zu spezifischen Anfragen bezüglich Ihrer Situation.",
       "dataSecurity": "Datenschutz und Sicherheit: \nIhre Privatsphäre ist uns sehr wichtig. Alle Gespräche werden vertraulich behandelt und Ihre Daten werden sicher verarbeitet. \nMehr Informationen zum Datenschutz finden Sie in unseren Datenschutzrichtlinien.",
+      "errorMessage": "Es ist ein Fehler bei der Kommunikation aufgetreten.",
       "header": "{{appName}} Chat Support",
+      "initialMessage": "Ihre Chatanfrage wurde gesendet. Die Hotline beantwortet Ihre Nachricht so schnell wie möglich.",
```
[the remark this answers is not in the document]
Comment (f1sh1918) [reply translations/translations.json:216]: thx for the hints. i used the suggestions from design and adjusted them.

**Comment by f1sh1918** [comment]
> Nice work! Tested on firefox, works perfectly!
> 
> Obviously not from this PR:
> 
> * the content behind the chat layover is scrollable
> * on small screens the send button is cut off at the bottom
>   image

okay i found the issue. This happens because of the label. I had to adjust the input container size depending on hasChat started or not. I have to define a fixed height here to be able to make the container scrollable. Its a bit ugly but there is no other way to do that afaik

File: web/src/components/Chat.tsx [hunk 15]
```diff
@@ -44,12 +44,12 @@ const StyledLoadingSpinner = styled(LoadingSpinner)`
   margin-top: 0;
 `
 
-const InputWrapper = styled.div`
-  height: ${dimensions.chatInputContainerHeight}px;
+const InputWrapper = styled.div<{ $height: number }>`
+  height: ${props => props.$height}px;
```
[the remark this answers is not in the document]
Comment (LeandraH) [reply web/src/components/Chat.tsx:48]: Ah, of course. Too bad!

**Pull request merged**